src/btree/bt_page.c

```diff
@@ -15,6 +15,575 @@ static int __inmem_row_int(WT_SESSION_IMPL *, WT_PAGE *, size_t *);
+    WT_INTL_INDEX_SET(ref->page, pindex);
+    __wt_cache_page_inmem_incr(session, ref->page, incr, false);
+
+    if (0) {
+err:
+        if (refs != NULL)
+            for (i = 0; i < (uint32_t)refs_entries; ++i)
+                __wt_free(session, refs[i]);
+    }
+    __wt_free(session, refs);
+    __wt_free(session, unpacked_deltas_merged);
+    if (unpacked_deltas != NULL) {
+        for (i = 0; i < (uint32_t)delta_size; ++i)
+            __wt_free(session, unpacked_deltas[i]);
+        __wt_free(session, unpacked_deltas);
+    }
+    __wt_free(session, delta_size_each);
+
+    return (ret);
+}
+
+/*
+ * __page_reconstruct_leaf_delta --
+ *     Reconstruct delta on a leaf page
+ */
+static int
+__page_reconstruct_leaf_delta(WT_SESSION_IMPL *session, WT_REF *ref, WT_ITEM *delta)
+{
+    WT_CELL_UNPACK_DELTA_LEAF unpack;
+    WT_CURSOR_BTREE cbt;
+    WT_DECL_RET;
+    WT_DELTA_HEADER *header;
+    WT_ITEM key, value;
+    WT_PAGE *page;
+    WT_ROW *rip;
+    WT_UPDATE *first_upd, *upd, *standard_value, *tombstone;
+    size_t size, tmp_size, total_size;
+
+    header = (WT_DELTA_HEADER *)delta->data;
+    tmp_size = total_size = 0;
+    page = ref->page;
+
+    WT_CLEAR(unpack);
+
+    __wt_btcur_init(session, &cbt);
+    __wt_btcur_open(&cbt);
+
+    WT_CELL_FOREACH_DELTA_LEAF(session, header, unpack)
+    {
+        key.data = unpack.key;
+        key.size = unpack.key_size;
+        upd = standard_value = tombstone = NULL;
+        size = 0;
+
+        /* Search the page and apply the modification. */
+        WT_ERR(__wt_row_search(&cbt, &key, true, ref, true, NULL));
+        /*
+         * We apply deltas from newest to oldest, ignore keys that have already got a delta update.
+         */
+        if (cbt.compare == 0) {
+            if (cbt.ins != NULL) {
+                if (cbt.ins->upd != NULL && F_ISSET(cbt.ins->upd, WT_UPDATE_RESTORED_FROM_DELTA))
+                    continue;
+            } else {
+                rip = &page->pg_row[cbt.slot];
+                first_upd = WT_ROW_UPDATE(page, rip);
+                if (first_upd != NULL && F_ISSET(first_upd, WT_UPDATE_RESTORED_FROM_DELTA))
+                    continue;
+            }
+        }
+
+        if (F_ISSET(&unpack, WT_DELTA_LEAF_IS_DELETE)) {
+            WT_ERR(__wt_upd_alloc_tombstone(session, &tombstone, &tmp_size));
+            F_SET(tombstone, WT_UPDATE_DURABLE | WT_UPDATE_RESTORED_FROM_DELTA);
+            size += tmp_size;
+            upd = tombstone;
+        } else {
+            value.data = unpack.value;
+            value.size = unpack.value_size;
+            WT_ERR(__wt_upd_alloc(session, &value, WT_UPDATE_STANDARD, &standard_value, &tmp_size));
+            standard_value->txnid = unpack.tw.start_txn;
+            standard_value->start_ts = unpack.tw.start_ts;
+            standard_value->durable_ts = unpack.tw.durable_start_ts;
+            F_SET(standard_value, WT_UPDATE_DURABLE | WT_UPDATE_RESTORED_FROM_DELTA);
```

Comment:
@tdlinh2712 We need to add the prepare id and prepare ts also here once this is merged.